core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.operator.exchange;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.log.Logger;
+import io.airlift.units.DataSize;
+import io.trino.operator.PartitionFunction;
+import io.trino.spi.Page;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class ScaleWriterPartitioningExchanger
```

Comment:
It's not clear to me what this class does. Could add a comment describing the intended behaviour?

Reply:
Added comments to this class. Also, Please take a look at the approach in this issue: https://github.com/trinodb/trino/issues/13379

Reply:
Nice description!